### ISSUE TITLE
GCC 8.2.1 needs -fno-gnu-unique to work around std::make_shared defect

### ISSUE THREAD

**Issue by abyss7** [MaskRay/ccls#363]
### Observed behavior

ccls crashes:

    14:01:31 indexer1         pipeline.cc:344 I parse gh/ch/dbms/src/Formats/FormatFactory.cpp
    14:01:31 preamble     sema_manager.cc:734 I create session for gh/ch/dbms/src/Formats/FormatFactory.cpp
    LLVM ERROR: out of memory
    LLVMSymbolizer: error reading file: No such file or directory
    #0 0x00007fe75b21d09e llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/lib64/libLLVM-7.so+0x9b409e)
    #1 0x00007fe75b21b554 llvm::sys::RunSignalHandlers() (/lib64/libLLVM-7.so+0x9b2554)
    #2 0x00007fe75b21b6d5 (/lib64/libLLVM-7.so+0x9b26d5)
    #3 0x00007fe75a523030 __restore_rt (/lib64/libpthread.so.0+0x13030)
    #4 0x00007fe75a38053f __GI_raise /usr/src/debug/glibc-2.28-60-g4d7af7815a/signal/../sysdeps/unix/sysv/linux/internal-signals.h:84:10
    #5 0x00007fe75a36a895 __GI_abort /usr/src/debug/glibc-2.28-60-g4d7af7815a/stdlib/abort.c:81:7
    #6 0x00007fe75b19fe34 llvm::report_bad_alloc_error(char const*, bool) (/lib64/libLLVM-7.so+0x936e34)
    #7 0x00007fe75b1c4bf1 llvm::SmallVectorBase::grow_pod(void*, unsigned long, unsigned long) (/lib64/libLLVM-7.so+0x95bbf1)
    #8 0x00007fe75b1fc0bd llvm::raw_svector_ostream::write_impl(char const*, unsigned long) (/lib64/libLLVM-7.so+0x9930bd)
    #9 0x00007fe75b1fcb97 llvm::raw_ostream::write(char const*, unsigned long) (/lib64/libLLVM-7.so+0x993b97)
    #10 0x00007fe75b1e895c llvm::Twine::print(llvm::raw_ostream&) const (/lib64/libLLVM-7.so+0x97f95c)
    #11 0x00007fe75b1e8b80 llvm::Twine::toVector(llvm::SmallVectorImpl<char>&) const (/lib64/libLLVM-7.so+0x97fb80)
    #12 0x00007fe75ea2c54f clang::vfs::InMemoryFileSystem::setCurrentWorkingDirectory(llvm::Twine const&) (/lib64/libclangBasic.so.7+0x1e254f)
    #13 0x00007fe75ea2a52e clang::vfs::OverlayFileSystem::pushOverlay(llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>) (/lib64/libclangBasic.so.7+0x1e052e)
    #14 0x00007fe75fe2c00e (/lib64/libclangFrontend.so.7+0x12500e)
    #15 0x00007fe75fe2c1fb clang::PrecompiledPreamble::setupPreambleStorage(clang::PrecompiledPreamble::PCHStorage const&, clang::PreprocessorOptions&, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>&) (/lib64/libclangFrontend.so.7+0x1251fb)
    #16 0x00007fe75fe2c696 clang::PrecompiledPreamble::configurePreamble(clang::PreambleBounds, clang::CompilerInvocation&, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>&, llvm::MemoryBuffer*) const (/lib64/libclangFrontend.so.7+0x125696)
    #17 0x0000000000483c54 (ccls+0x483c54)
    #18 0x000000000047fed8 (ccls+0x47fed8)
    #19 0x00007fe75a51858e start_thread /usr/src/debug/glibc-2.28-60-g4d7af7815a/nptl/pthread_create.c:487:7
    #20 0x00007fe75a4456a3 __GI___clone /usr/src/debug/glibc-2.28-60-g4d7af7815a/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97:0
    ccls server connection was closed
    [Error - 2:01:40 PM] Connection to server got closed. Server will not be restarted.

### Expected behavior

ccls should reindex new contents.

### Steps to reproduce

1. git checkout new revision.
2. Open any cpp changed file in vscode.
3. Try to use index in any way: hover cursor, find references, etc.

### System information

* ccls version (`git describe --tags --long`): 0.20181225.9-35-g1976fec5
* clang version: clang version 7.0.1 (tags/RELEASE_701/final)
* OS: Fedora 29
* Editor: vscode
* Language client (and version): vscode-ccls 0.1.26

**Comment by MaskRay** [comment]
Have you tried building ccls with clang 8? If your distribution doesn't provide it, consider using prebuilt archives on http://releases.llvm.org/

> `#17 0x0000000000483c54 (ccls+0x483c54)`
> `#18 0x000000000047fed8 (ccls+0x47fed8)`

A `-DCMAKE_BUILD_TYPE=Debug` build will be helpful

**Comment by rompy** [comment]
Had an similar problem (crash whenever a file was indexed), which was fixed by using `-DCMAKE_CXX_COMPILER=clang`

**Comment by 0xC0FFEE** [comment]
> Had an similar problem (crash whenever a file was indexed), which was fixed by using `-DCMAKE_CXX_COMPILER=clang`

I had this bug too and discovered by accident that using clang to compile ccls fixes this. Tested with clang 8.0.0 and gcc 8.2.1.

Here is the (partial) backtrace of the core dump:

```
#0  0x00005596fad06257 _ZNSt23_Sp_counted_ptr_inplaceIN4llvm3sys2fs6detail12DirIterStateESaIS4_ELN9__gnu_cxx12_Lock_policyE2EE14_M_get_deleterERKSt9type_info (ccls)
#1  0x00007f352616f77b n/a (libLLVM-8.so)
#2  0x00007f352a0903b1 n/a (libclangDriver.so.8)
#3  0x00007f352a091360 n/a (libclangDriver.so.8)
#4  0x00007f352a0aa015 n/a (libclangDriver.so.8)
#5  0x00007f3529fdbb7b _ZNK5clang6driver6Driver12getToolChainERKN4llvm3opt7ArgListERKNS2_6TripleE (libclangDriver.so.8)
#6  0x00007f3529fe76a4 _ZN5clang6driver6Driver16BuildCompilationEN4llvm8ArrayRefIPKcEE (libclangDriver.so.8)
#7  0x00007f352b0357a3 _ZN5clang31createInvocationFromCommandLineEN4llvm8ArrayRefIPKcEENS0_18IntrusiveRefCntPtrINS_17DiagnosticsEngineEEENS5_INS0_3vfs10FileSystemEEE (libclangFrontend.so.8)
#8  0x00005596fad04177 n/a (ccls)
#9  0x00005596fad1ba15 n/a (ccls)
#10 0x00005596fad393e8 n/a (ccls)
#11 0x00005596fad3ca76 n/a (ccls)
#12 0x00005596fadd60e5 n/a (ccls)
#13 0x00007f35255f4a9d start_thread (libpthread.so.0)
#14 0x00007f3525524b23 __clone (libc.so.6)
```

**Comment by yssource** [comment]
the similar problem happen to me also. 
- archlinuxcn/ccls-git 20190329-2
- emacs27

``` bash
clang version 8.0.0 (tags/RELEASE_800/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1
Found candidate GCC installation: /usr/lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/8.2.1
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
```

ccls server wont start

**Comment by yssource** [comment]
``` bash
10:07:06 ccls           initialize.cc:271 I initialize in directory /tmp with uri file:///tmp/
10:07:06 ccls           initialize.cc:294 I initializationOptions: {"compilationDatabaseCommand":"","compilationDatabaseDirectory":"","cache":{"directory":"/home/jimmy/.emacs.d/.cache/lsp-ccls","format":"binary","hierarchicalPath":false,"retainInMemory":2},"capabilities":{"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"}","moreTriggerCharacter":[]},"foldingRangeProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}},"clang":{"excludeArgs":[],"extraArgs":[],"pathMappings":[],"resourceDir":""},"client":{"hierarchicalDocumentSymbolSupport":true,"linkSupport":true,"snippetSupport":true},"codeLens":{"localVariables":true},"completion":{"caseSensitivity":2,"detailedLabel":true,"dropOldRequests":true,"duplicateOptional":true,"filterAndSort":true,"include":{"blacklist":[],"maxPathSize":30,"suffixWhitelist":[".h",".hpp",".hh",".inc"],"whitelist":[]},"maxNum":100},"diagnostics":{"blacklist":[],"onChange":1000,"onOpen":0,"onSave":0,"spellChecking":true,"whitelist":[]},"highlight":{"largeFileSize":2097152,"lsRanges":false,"blacklist":[],"whitelist":[]},"index":{"blacklist":[],"comments":2,"initialNoLinkage":false,"initialBlacklist":[],"initialWhitelist":[],"maxInitializerLines":5,"multiVersion":0,"multiVersionBlacklist":[],"multiVersionWhitelist":[],"name":{"suppressUnwrittenScope":false},"onChange":false,"parametersInDeclarations":true,"threads":0,"trackDependency":2,"whitelist":[]},"request":{"timeout":5000},"session":{"maxNum":10},"workspaceSymbol":{"caseSensitivity":1,"maxNum":1000,"sort":true},"xref":{"maxNum":2000}}
10:07:06 ccls           initialize.cc:323 I use -resource-dir=/usr/lib/clang/8.0.0
10:07:06 ccls           initialize.cc:356 I workspace folder: /tmp/
10:07:06 ccls           initialize.cc:381 I start 4 indexers
10:07:06 ccls           initialize.cc:389 I dispatch initial index requests
10:07:06 ccls             pipeline.cc:478 I loaded project. Refresh semantic highlight for all working file.
10:07:06 preamble     sema_manager.cc:733 I create session for /tmp/hello/hello.cpp
10:07:06 indexer3         pipeline.cc:344 I parse /tmp/hello/hello.cpp

Process ccls stderr finished

```

**Comment by yssource** [comment]
- git clone --recursive https://github.com/MaskRay/ccls
- cd ccls
- cmake -H. -BRelease -G Ninja -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=clang++
- ninja -C Release

It solves my problem.

**Comment by wbthomason** [comment]
@yssource I can confirm that this also fixed the issue for me on Arch; by adding `-DCMAKE_CXX_COMPILER=clang++` to the normal AUR PKGBUILD I was able to get a build of `ccls` that does not crash. It seems like that should be added to the official PKGBUILD.

@yhamdoud, you may also be interested in this.

**Comment by madscientist** [comment]
Interesting: I have ccls built with GCC 8.1.0 and I don't see this issue.  This GCC is one I built myself so maybe there's some configuration of the Fedora version of GCC that is causing this problem?

**Comment by yhamdoud** [comment]
> @yssource I can confirm that this also fixed the issue for me on Arch; by adding `-DCMAKE_CXX_COMPILER=clang++` to the normal AUR PKGBUILD I was able to get a build of `ccls` that does not crash. It seems like that should be added to the official PKGBUILD.
> 
> @yhamdoud, you may also be interested in this.

Thanks, this also worked for me.

**Comment by abyss7** [comment]
> Have you tried building ccls with clang 8? If your distribution doesn't provide it, consider using prebuilt archives on http://releases.llvm.org/
> 
> > `#17 0x0000000000483c54 (ccls+0x483c54)`
> > `#18 0x000000000047fed8 (ccls+0x47fed8)`
> 
> A `-DCMAKE_BUILD_TYPE=Debug` build will be helpful

I built it with Clang 8 and can't reproduce the issue anymore. Can you clarify, how the compiler may help in this situation?

**Comment by MaskRay** [comment]
I'll jot down some notes here and keep updating this comment.

If you build ccls with extra/clang

% readelf -W --dyn-syms ~/abs/ccls-git/pkg.good/ccls-git/usr/bin/ccls | grep _Sp_make_shared_tag
   412: 00000000001214a0    16 OBJECT  WEAK   DEFAULT   15 _ZZNSt19_Sp_make_shared_tag5_S_tiEvE5__tag
   507: 0000000000121486    24 OBJECT  WEAK   DEFAULT   15 _ZTSSt19_Sp_make_shared_tag

If you build ccls with gcc 8.2.1

```
% readelf -W --dyn-syms ~/abs/ccls-git/pkg.bad/ccls-git/usr/bin/ccls | grep _Sp_make_shared_tag
   370: 0000000000140608    16 OBJECT  UNIQUE DEFAULT   13 _ZZNSt19_Sp_make_shared_tag5_S_tiEvE5__tag
   386: 0000000000140530    24 OBJECT  WEAK   DEFAULT   13 _ZTSSt19_Sp_make_shared_tag
```

For `std::shared_ptr` to work correctly, all references to the symbol `_ZTSSt19_Sp_make_shared_tag` must resolve to the main executable.

Now the issue is that for GCC 8.2.1 compiled ccls, the `_ZTSSt19_Sp_make_shared_tag` reference from `libLLVM-8.so` unexpectedly resolves to `libiccuc.so.64`. I haven't figured out the cause. Not sure if it is a glibc bug or the GCC compiled object files do something special.

```
% readelf --dyn-syms /usr/lib/libLLVM-8.so G _Sp_make_shared_tag
  3677: 0000000002e2d4e8    16 OBJECT  UNIQUE DEFAULT   12 _ZZNSt19_Sp_make_shared_tag5_S_tiEvE5__tag@@LLVM_8
 13290: 0000000002e2d290    24 OBJECT  WEAK   DEFAULT   12 _ZTSSt19_Sp_make_shared_tag@@LLVM_8

% readelf --dyn-syms /usr/lib/libicuuc.so.64 G _Sp_make_shared_tag
   351: 0000000000140950    16 OBJECT  UNIQUE DEFAULT   11 _ZZNSt19_Sp_make_shared_tag5_S_tiEvE5__tag
  1508: 0000000000140890    24 OBJECT  WEAK   DEFAULT   11 _ZTSSt19_Sp_make_shared_tag
```

**Comment by bastianbeischer** [comment]
Hm, if I understand the above correctly, then ccls built with GCC 8.3.0 should work fine, right? It includes the fix for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=88782. But I'm still seeing the same issues as the OP... Am I missing something?

**Comment by bastianbeischer** [comment]
Ah - now I understand: LLVM also has to be built with the new GCC version. Problem is fixed with Arch llvm package llvm-8.0.0-2. Thanks!
